grammars/silver/compiler/extension/records/Records.sv.md

```diff
@@ -0,0 +1,370 @@
+# Records
+
+```silver imports
+grammar silver:compiler:extension:records;
+
+import silver:compiler:definition:core;
+import silver:compiler:definition:env;
+import silver:compiler:definition:type;
+import silver:compiler:definition:type:syntax;
+```
+
+The records extension is meant to codify a common pattern of defining product types that generates efficient code and has nice usability.
+
+## Declaration
+
+A record definition starts with the `record` keyword.
+
+```silver
+terminal RecordKwd_t 'record' lexer classes {KEYWORD};
+```
+
+A record definition looks basically like a nonterminal declaration followed by a list of fields.
+This forwards to a nonterminal, a single production with no children, and an annotation that occurs on the nonterminal for each field.
+
+```silver
+concrete production recordDecl
+this::AGDcl ::= quals::NTDeclQualifiers  'record'  name::IdUpper_t
+  modifiers::NonterminalModifiers  '{'  fields::RecordFields0  '}'  ';'
+{
+  this.unparse =
+    s"${quals.unparse}record ${name.lexeme} ${modifiers.unparse} {${fields.unparse}};";
+
+  local ntDcl::AGDcl =
+    nonterminalDcl(quals, 'nonterminal',
+      nameIdUpper(name, location=this.location),
+      botlNone(location=this.location), modifiers, ';', location=this.location);
+  local ntType::TypeExpr =
+    nominalTypeExpr(qNameTypeId(name, location=this.location),
+      location=this.location);
+
+  -- PARTIAL: head() and tail() are safe because names are nonempty.
+  local nameChars::[String] = explode("", name.lexeme);
+  local prodName::IdLower_t = terminal(IdLower_t,
+    implode("", toLower(head(nameChars)) :: tail(nameChars)),
+    name.location);
+
+  local prodLHS::ProductionLHS = productionLHS(
+    nameIdLower(terminal(IdLower_t, "this"), location=this.location), '::',
+    ntType, location=this.location);
+  local prodSig::ProductionSignature =
+    productionSignatureNoCL(prodLHS, '::=',
+      productionRHSNil(location=this.location), location=this.location);
+  local prodDcl::AGDcl =
+    productionDcl('abstract', 'production',
+      nameIdLower(prodName, location=this.location), prodSig, 
+      productionBody('{', productionStmtsNil(location=this.location), '}',
+        location=this.location),
+      location=this.location);
+
+  forwards to appendAGDcl(ntDcl,
+    appendAGDcl(prodDcl, fields.recordFieldAGDcl, location=this.location),
+    location=this.location);
+}
+```
+
+### `without declaring` Nonterminal Modifier
```

Comment:
Uhhhh I really don't like this syntax.  I would strongly prefer to just look at fixing the env to allow us to avoid duplicating annotation declarations.